sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java

```diff
@@ -522,4 +651,393 @@ public Read withMaxCapacityPerShard(Integer maxCapacity) {
+        FluentBackoff backoff =
+            (retryConfig != null)
+                ? retryConfig
+                    .toThrottlingBackoff()
+                    .withMaxRetries(max(MIN_PARTIAL_RETRIES, retryConfig.numRetries()))
+                : FluentBackoff.DEFAULT.withMaxRetries(MIN_PARTIAL_RETRIES);
+        KinesisAsyncClient client = clients.retain(spec.kinesisClientProvider(), retryConfig);
+        handler = new AsyncPutRecordsHandler(client, spec.concurrentRequests(), backoff);
+      }
+
+      @StartBundle
+      public void startBundle() {
+        requestEntries = new ArrayList<>();
+        requestBytes = 0;
+        handler.reset();
+      }
+
+      @ProcessElement
+      public final void processElement(ProcessContext c) throws Throwable {
+        handler.checkForAsyncFailure();
+        T record = c.element();
+        byte[] data = spec.serializer().apply(record);
+        String pk = spec.partitioner().getPartitionKey(record);
+        String ehk = spec.partitioner().getExplicitHashKey(record);
+
+        validatePartitionKey(pk);
+        if (ehk != null) {
+          validateExplicitHashKey(ehk);
+        }
+        processElement(pk, ehk, data);
+      }
+
+      protected void processElement(
+          String partitionKey, @Nullable String explicitHashKey, byte[] data) throws Throwable {
+        addRequestEntry(
+            PutRecordsRequestEntry.builder()
+                .data(SdkBytes.fromByteArrayUnsafe(data))
+                .partitionKey(partitionKey)
+                .explicitHashKey(explicitHashKey)
+                .build(),
+            spec.batchMaxRecords());
+      }
+
+      protected final void addRequestEntry(PutRecordsRequestEntry entry, int maxEntries)
+          throws Throwable {
+        int entryBytes =
+            entry.partitionKey().getBytes(UTF_8).length + entry.data().asByteArrayUnsafe().length;
+        // check first if new record can still be added to batch, flush otherwise
+        if (requestBytes + entryBytes > spec.batchMaxBytes()) {
+          flushEntries();
+        }
+        // the entry might exceed batchMaxBytes, but try anyways
+        requestEntries.add(entry);
+        requestBytes += entryBytes;
+        // flush once batch size limit is reached
+        if (requestEntries.size() >= min(maxEntries, spec.batchMaxRecords())) {
+          flushEntries();
+        }
+      }
+
+      private void flushEntries() throws Throwable {
+        if (!handler.hasErrored() && !requestEntries.isEmpty()) {
+          // Swap lists, luckily no need to synchronize
+          List<PutRecordsRequestEntry> recordsToWrite = requestEntries;
+          requestEntries = new ArrayList<>();
+          requestBytes = 0;
+          handler.putRecords(spec.streamName(), recordsToWrite);
+        }
+      }
+
+      @FinishBundle
+      public void finishBundle() throws Throwable {
+        flushEntries();
+        handler.waitForCompletion();
+      }
+
+      @Teardown
```

Comment:
`@Teardown` is a "best effort" to be called. Is it ok to rely on it in this case?

Reply:
ping

Reply:
yes 👍